File: Cdb/src/main/python/IasCdb/TextFileType.py
from pathlib import Path
from enum import Enum
import os
from typing import Union

class FileType(Enum):
    YAML=1,
    JSON=2

class TextFileType:

    # Supported file types and their extension
    SUPPORTED_TYPES = { FileType.YAML:".yaml", FileType.JSON:".json"}

    @classmethod
    def from_file(cls, file_name: str) -> Union[FileType, None]:
        """
        Get the type of the file from the extension of the passed file name

        Args:
            file_name: the file name
        Returns:
            The type of the file (belongs to FileType) 
            or None if the type has not been recognized
        """
        if not file_name:
            raise ValueError("Invalid empty file name")
        
        # Get the extension of the file
        pos = file_name.rfind('.')
        if pos == -1:
            raise ValueError("File has no extension: " + file_name)

        ext = file_name[pos:].lower()
        if ext == TextFileType.SUPPORTED_TYPES[FileType.JSON]:
            return FileType.JSON
        elif ext == TextFileType.SUPPORTED_TYPES[FileType.YAML]:
            return FileType.YAML
        else:
            return None

    @classmethod
    def get_cdb_type(cls, folder: str) -> Union[FileType, None]:
        """
        Get the type of the CDB from the passed folder.
        It checks the content of the CDB folder looking for a file named ias.ext

        Args:
            folder: The parent folder of the CDB
        Returns:
            the type of the text files in the CDB folder,
            or empty if cannot recognize the type of the CDB of the text files
        """
        if not folder:
            raise ValueError(f"Invalid empty folder")

        cdb_folder = folder+"/CDB"
        if not os.path.isdir(cdb_folder):
            raise ValueError(f"The folder {cdb_folder} is unreadable or does not exist")

        # Look for ias.*: there should be only one
        files = [f.name for f in cdb_folder.iterdir() if f.name.startswith("ias.")]
        if len(files) != 1:
            return None
        else:
            return cls.from_file(files[0])

File: Cdb/src/main/java/org/eso/ias/cdb/structuredtext/TextFileType.java
package org.eso.ias.cdb.structuredtext;

import java.io.File;
import java.io.FileFilter;
import java.io.FilenameFilter;
import java.nio.file.DirectoryStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Objects;
import java.util.Optional;

/**
 * Formatted text file supported by the CDB
 */
public enum TextFileType {
    JSON(".json"),
    YAML(".yaml");

    /** The extension of the supported type */
    public final String ext;

    private TextFileType(String ext) {
        this.ext=ext;
    }

    /**
     * Return the type of the CDB depending on the extension of
     * the passed file name.
     *
     * @param fileName the file name
     * @return the type corresponding to this file name
     *         or empty if the type has not been recognized
     */
    public static Optional<TextFileType> fromFile(String fileName) {
        Objects.requireNonNull(fileName);
        if (fileName.isEmpty()) {
           throw new IllegalArgumentException("Invalid empty file name");
        }
        // Get the extension of the file
        int pos=fileName.lastIndexOf('.');
        if (pos==-1) {
           throw new IllegalArgumentException("File has no extension: "+fileName);
        }
        String ext=fileName.substring(pos);
        if (ext.compareToIgnoreCase(TextFileType.JSON.ext)==0) {
            return Optional.of(TextFileType.JSON);
        } else if (ext.compareToIgnoreCase(TextFileType.YAML.ext)==0) {
            return Optional.of(TextFileType.YAML);
        } else {
            return Optional.empty();
        }
    }

    /**
     * Get the type of the CDB from the passed folder.
     *
     * It checks the content of the CDB folder looking for
     * a file named ias.<literal>.<extension></literal>.
     *
     * @param folder The parent folder of the CDB
     * @return the type of the text files in the CDB folder,
     *         or empty if cannot recognize the type of the CDB of the text files
     * @throws IllegalArgumentException if the passed parameter does not correspond to the parent folder
     *                                  of a CDB
     */
    public static Optional<TextFileType> getCdbType(File folder) {
        Objects.requireNonNull(folder);
        if (!folder.isDirectory() || !folder.canRead()) {
            throw new IllegalArgumentException("Not a folder or unreadable: "+folder.getAbsolutePath());
        }
        File cdbFolder = new File(folder.getAbsolutePath()+File.separator+"CDB");
        if (!cdbFolder.isDirectory() || !cdbFolder.canRead()) {
            throw new IllegalArgumentException("The folder "+cdbFolder.getAbsolutePath()+" is unreadable or does not exist");
        }
        // Look for ias.*: there should be only one
        FilenameFilter filter = new FilenameFilter() {
            @Override
            public boolean accept(File file, String s) {
                return s.startsWith("ias.");
            }
        };
        File[] files = cdbFolder.listFiles(filter);
        if (files.length!=1) {
            return Optional.empty();
        } else {
            return fromFile(files[0].getName());
        }
    }
}
